clojure-north.org

```diff
@@ -0,0 +1,360 @@
+  assume that the Kafka streams DSL transformations are always correct. The
+  problem is comes when you want to validate behavior of a topology as data
+  flows through it. Using spec to validate data within the boundaries of the
+  transformation functions goes a long way but does not quite cover everything.
+
+  In an ideal world all your tests should be free of external dependencies and
+  the topology level tests we should cover everything. Unfortunately that's not
+  the case for Kafka steams apps because there is a lot more to validating a
+  topology. Mainly you also want to validate that data coming in an out of
+  topology nodes is in the right shape and using the correct schema. At Funding
+  Circle we use AVRO to define the schema of most of our topics. It is super
+  common to run into bugs where data in the topology looks good but the does not
+  match a schema of the other way around.
+
+  Fortunately we now have a few ways we learned to deal with this issues.
+
+  For the schema vs (EDN) data in our Topologies rely on Clojure specs. Having
+  specs to describe the data is great but we also use it to generate the AVRO
+  schemas we are going to write to. We have spec-to-avro compiler which we use
+  to publish and update the AVRO schemas our topics support and in our tests we
+  rely on generators to produce any dummy data we want to test with in addition
+  to being able to do generative testing to achieve the ultimate level of
+  confidence in our tests. Recently we even started to create different sets of
+  specs for reading and writing operations as the schema constraints of writing
+  to a topic and reading from a topic drift over time as a schema evolves. This
+  is particularly more painful because of our use of [[https://docs.confluent.io/current/schema-registry/avro.html][AVRO]].
+
+  For validating the behavior of a topology (integration testing) we now rely on
+  [[https://github.com/FundingCircle/jackdaw/blob/master/doc/test-machine.md][Test Machine]]. This is a newer addition to the Jackdaw library that uses the
+  ~TopologyTestDriver~ that appeared in Kafka version 2.4.0. It allows us to
+  create a fixture that allows us to run a topology without the need to run a
+  Kafka. This is actually a very interesting subject but it is a bit out of
+  scope of this presentation. I'll add a link to the references where you can
+  learn more on this subject. The ~TopologyTestDriver~ covers most of our
+  integration testing needs but sometimes we want to test the behavior of more
+  than one topology at the time or the test the interaction of topology with the
+  schema registry and a mocked service. For this scenarios we Test Machine
+  actually supports multiple drivers. Allowing us to setup tests that can run
+  against a Kafka or Kafka rest proxy. This is a super powerful feature of
+  Jackdaw I suggest you learn more about.
+
+  Finally, we arrive to the top of the testing pyramid. The "god" level tests;
```

Comment:
If you weren't planning on it, I'd suggest having visuals that step through a testing pyramid to help the audience and you be on the same page.

Reply:
Also, maybe some sort of thing like:

"So, we get to the top of the testing pyramid and here is where we have our expensive tests. The ones that test the whole system. The ones that when they fail, they leave you wondering 'what happened?' I always connect these types of test to my personal health story"

(obviously, feel free to ignore anything I'm suggesting, I won't be sad)